src/libs/safe-path/src/pinned_path_buf.rs

```diff
@@ -0,0 +1,380 @@
+///
+/// Note:
+/// - Though the value of [PinnedPathBuf::target()] never changes, but the filesystem object
+///   referenced by it may be changed. Only the filesystem object referenced by
+///   `PinnedPathBuf::as_path()` is stable.
+/// - Once the `PinnedPathBuf` object gets dropped, the Path returned by `PinnedPathBuf::as_path()`
+///   gets invalid.
+///
+/// With normal [PathBuf], there's a race window for attackers between time to validate a path and
+/// time to use the path. An attacker may maliciously change filesystem object referenced by the
+/// path by symlinks to compose an attack.
+///
+/// The [PinnedPathBuf] is introduced to protect from such attacks, by using the `/proc/self/fd/xxx`
+/// files on Linux. The `/proc/self/fd/xxx` file on linux is a symlink to the real target
+/// corresponding to the process's file descriptor `xxx`. And the symlink will be kept stable until
+/// the file descriptor has been closed. Combined with `O_PATH`, a safe version of `PathBuf` could
+/// be built by:
+/// - Generate a safe path from `root` and `path` by using [crate::scoped_join()].
+/// - Open the safe path with O_PATH | O_CLOEXEC flags, say it's assigned `fd_num`.
+/// - Read the symlink target of `/proc/self/fd/fd_num`.
+/// - Compare the symlink target with the safe path, it's safe if these two paths equal.
+/// - Use the prof file path as a safe PathBuf.
+/// - Close the `fd_num` when dropping the `PinnedPathBuf` object.
+#[derive(Debug)]
+pub struct PinnedPathBuf {
+    handle: File,
+    path: PathBuf,
+    target: PathBuf,
+}
+
+impl PinnedPathBuf {
+    /// Create a `PinnedPathBuf` from `root` and `path`.
+    ///
+    /// The `path` must be a subdirectory of `root`, otherwise error will be returned.
+    pub fn new<R: AsRef<Path>, U: AsRef<Path>>(root: R, path: U) -> Result<Self> {
+        let path = scoped_join(root, path)?;
+        Self::from_path(path)
+    }
+
+    /// Create a `PinnedPathBuf` from `path`.
+    ///
+    /// If the resolved value of `path` doesn't equal to `path`, an error will be returned.
+    pub fn from_path<P: AsRef<Path>>(orig_path: P) -> Result<Self> {
+        let orig_path = orig_path.as_ref();
+        let handle = Self::open_by_path(orig_path)?;
+        Self::new_from_file(handle, orig_path)
+    }
+
+    /// Open a direct child of the `PinnedPathBuf` object.
+    pub fn open_child(&self, comp: &str) -> Result<Self> {
+        if comp.contains(std::path::is_separator) {
+            return Err(Error::new(
+                ErrorKind::Other,
+                format!("Path component {} is invalid", comp),
+            ));
+        }
+
+        let handle = self.openat_by_path(comp)?;
+        Self::new_from_file(handle, self.target.join(comp))
+    }
+
+    /// Try to clone the `PinnedPathBuf` object.
+    pub fn try_clone(&self) -> Result<Self> {
+        let fd = unsafe { libc::dup(self.path_fd()) };
+        if fd < 0 {
+            Err(Error::last_os_error())
```

Comment:
This technique of obtaining the error looks like it might be racy to me. What if the caller was an `async` task which was moved to a different O/S thread by the async runtime? Wouldn't that invalidate querying `errno` (which is thread-local)?